### PR TITLE
HPCC-18013 Add method to get RtlRecord from IOutputMetaData

### PR DESCRIPTION
**Pull request by richardkchapman**
Also fixes various internal errors encountered when using VIRTUAL RECORD to
signify passing just the row metadata to external functions.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
Pending - will update

### CODE REVIEW

File: common/thorhelper/thorcommon.ipp [hunk 1]
```diff
@@ -540,6 +539,7 @@ public:
     {
         return original->queryChildMeta(i);
     }
+    virtual const RtlRecord *queryRecordAccessor() const { return original->queryRecordAccessor(); }
```
Comment (ghalliday) [inline common/thorhelper/thorcommon.ipp]: Needs a boolean parameter

File: rtl/include/eclhelper.hpp [hunk 2]
```diff
@@ -473,6 +475,7 @@ interface IOutputMetaData : public IRecordSize
     virtual void process(const byte * self, IFieldProcessor & target, unsigned from, unsigned to) {}            // from and to are *hints* for the range of fields to call through with
     virtual void walkIndirectMembers(const byte * self, IIndirectMemberVisitor & visitor) = 0;
     virtual IOutputMetaData * queryChildMeta(unsigned i) = 0;
+    virtual const RtlRecord *queryRecordAccessor(bool expand=true) const { return NULL; }
```
Comment (ghalliday) [inline rtl/include/eclhelper.hpp]: possibly better without a default parameter so it is clear which one is being used.

File: rtl/eclrtl/rtlrecord.cpp [hunk 3]
```diff
@@ -260,3 +352,132 @@ RtlDynRow::~RtlDynRow()
     delete [] variableOffsets;
 }
 
+//-----------------
+
+const RtlRecord *setupRecordAccessor(const COutputMetaData &meta, std::atomic<const RtlRecord *> &aRecordAccessor)
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: expand should be passed.

File: rtl/eclrtl/rtlrecord.cpp [hunk 4]
```diff
@@ -260,3 +352,132 @@ RtlDynRow::~RtlDynRow()
...
+}
+
+const RtlRecord *COutputMetaData::queryRecordAccessor(bool expand) const
+{
+    // NOTE: the recordAccessor field cannot be declared as atomic, since the class definition is included in generated
+    // code which is not (yet) compiled using C++11. If that changes then the reinterpret_cast can be removed.
+    std::atomic<const RtlRecord *> &aRecordAccessor = reinterpret_cast<std::atomic<const RtlRecord *> &>(recordAccessor[expand]);
+    const RtlRecord *useAccessor = aRecordAccessor.load(std::memory_order_relaxed);
+    if (!useAccessor)
+        useAccessor = setupRecordAccessor(*this, aRecordAccessor);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: pass expand

File: rtl/include/eclhelper.hpp (hunk 2 above)
Comment (richardkchapman) [inline rtl/include/eclhelper.hpp]: probably true. Will change

File: common/thorhelper/thorcommon.ipp (hunk 1 above)
Comment (richardkchapman) [inline common/thorhelper/thorcommon.ipp]: Indeed. Will change.

File: rtl/eclrtl/rtlrecord.cpp (hunk 3 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.cpp]: Fixed

File: rtl/eclrtl/rtlrecord.cpp (hunk 4 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.cpp]: Fixed

**Comment by AttilaVamos** [comment]
Some additional info about the build problem:
```
[ 92%] Linking CXX executable ../../Release/bin/thormaster_lcr
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/thorslave_lcr] Error 1
make[1]: *** [thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/thormaster_lcr] Error 1
```

**Comment by richardkchapman** [comment]
@AttilaVamos I think my latest push should have fixed the build error
@ghalliday  I think I fixed the issues you mention are fixed

**Comment by AttilaVamos** [comment]
I'm checking, what's going on.

**Comment by AttilaVamos** [comment]
Smoketest on n Ubuntu 16.04 has same problem:
[100%] Linking CXX executable ../../Release/bin/thorslave_lcr
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
thorlcr/slave/CMakeFiles/thorslave_lcr.dir/build.make:206: recipe for target 'Release/bin/thorslave_lcr' failed
make[2]: *** [Release/bin/thorslave_lcr] Error 1
CMakeFiles/Makefile2:14565: recipe for target 'thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all' failed
make[1]: *** [thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Linking CXX executable ../../Release/bin/thormaster_lcr
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
thorlcr/master/CMakeFiles/thormaster_lcr.dir/build.make:209: recipe for target 'Release/bin/thormaster_lcr' failed
make[2]: *** [Release/bin/thormaster_lcr] Error 1

**Comment by AttilaVamos** [comment]
Sorry that was the result of previous commit. Process started on the latest one.

**Comment by AttilaVamos** [comment]
Unfortunately there is same problem with latest commit:
```
[100%] Linking CXX executable ../../Release/bin/thorslave_lcr
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
thorlcr/slave/CMakeFiles/thorslave_lcr.dir/build.make:206: recipe for target 'Release/bin/thorslave_lcr' failed
make[2]: *** [Release/bin/thorslave_lcr] Error 1
CMakeFiles/Makefile2:14565: recipe for target 'thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all' failed
make[1]: *** [thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Linking CXX executable ../../Release/bin/thormaster_lcr
../../Release/libs/libgraph_lcr.so: undefined reference to `vtable for CFixedOutputMetaData'
collect2: error: ld returned 1 exit status
thorlcr/master/CMakeFiles/thormaster_lcr.dir/build.make:209: recipe for target 'Release/bin/thormaster_lcr' failed
make[2]: *** [Release/bin/thormaster_lcr] Error 1
CMakeFiles/Makefile2:14341: recipe for target 'thorlcr/master/CMakeFiles/thormaster_lcr.dir/all' failed
make[1]: *** [thorlcr/master/CMakeFiles/thormaster_lcr.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
Error(s): 9
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/thorslave_lcr] Error 1
make[1]: *** [thorlcr/slave/CMakeFiles/thorslave_lcr.dir/all] Error 2
collect2: error: ld returned 1 exit status
make[2]: *** [Release/bin/thormaster_lcr] Error 1
make[1]: *** [thorlcr/master/CMakeFiles/thormaster_lcr.dir/all] Error 2
make: *** [all] Error 2
```

File: common/thorhelper/thorfile.cpp [hunk 5]
```diff
@@ -25,7 +25,7 @@
 #include "rtlfield.hpp"
 #include "rtlds_imp.hpp"
 
-namespace thorfile {
+namespace thorfile {  // Make sure we can't clash with generated versions or version check mechanism fails.
```
Comment (ghalliday) [inline common/thorhelper/thorfile.cpp]: trivial: for consistency we should have the '{' on the next line (same for examples below)

File: rtl/eclrtl/rtlrecord.cpp [hunk 6]
```diff
@@ -219,6 +286,31 @@ size32_t RtlRecord::getMinRecordSize() const
     return minSize;
 }
 
+const FieldNameToFieldNumMap *setupNameMap(const RtlFieldInfo * const * fields, bool expand, std::atomic<const FieldNameToFieldNumMap *> &aNameMap)
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: trivial: also make this static

File: rtl/eclrtl/rtlrecord.cpp [hunk 7]
```diff
@@ -260,3 +352,132 @@ RtlDynRow::~RtlDynRow()
...
+        delete lRecordAccessor;
+        return expected;  // has been updated to the value set by other thread
+    }
+}
+
+const RtlRecord *COutputMetaData::queryRecordAccessor(bool expand) const
+{
+    // NOTE: the recordAccessor field cannot be declared as atomic, since the class definition is included in generated
+    // code which is not (yet) compiled using C++11. If that changes then the reinterpret_cast can be removed.
+    std::atomic<const RtlRecord *> &aRecordAccessor = reinterpret_cast<std::atomic<const RtlRecord *> &>(recordAccessor[expand]);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: I wondered if index[bool] was safe, but it looks like it is - explicitly defined to cast fasle->0 true->1

File: rtl/eclrtl/rtlrecord.hpp [hunk 8]
```diff
@@ -25,22 +25,189 @@
...
+public:
+    inline CXmlToRowTransformer(unsigned _activityId) : CGlobalHelperClass(_activityId) {}
+    RTLIMPLEMENT_IINTERFACE
+
+};
+
+//---------------------------------------------------------------------------
+// Record size handling.
+
+class ThorHelper : public RtlCInterface
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.hpp]: unrelated.  I think this class is unused.

File: rtl/eclrtl/rtlrecord.hpp [hunk 9]
```diff
@@ -25,22 +25,189 @@
...
-// Probably convert to an interface with various concrete implementations for varing degrees of complexity
+// Probably convert to an interface with various concrete implementations for varying degrees of complexity
 //
 // Complications:
 // * ifblocks
 // * nested records.
 // * alien data types
 //
 
+interface FieldNameToFieldNumMap;
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.hpp]: should be a class rather than interface - windows will complain

File: rtl/eclrtl/rtlrecord.hpp [hunk 10]
```diff
@@ -148,4 +318,92 @@ class ECLRTL_API RtlRecordSize : public IRecordSize, public RtlCInterface
...
 };
 
+class CSourceRowPrefetcher;
+
+class ECLRTL_API COutputMetaData : implements IOutputMetaData, public RtlCInterface
+{
+public:
+    RTLIMPLEMENT_IINTERFACE
+    COutputMetaData() { recordAccessor[0] = recordAccessor[1] = NULL; }
+    ~COutputMetaData() { delete recordAccessor[0]; delete recordAccessor[1]; }
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.hpp]: Possibly better out of line - no need for these functions to be inline any more.

File: rtl/eclrtl/rtlrecord.hpp [hunk 11]
```diff
@@ -148,4 +318,92 @@ class ECLRTL_API RtlRecordSize : public IRecordSize, public RtlCInterface
...
+
+    virtual void toXML(const byte * self, IXmlWriter & out) {
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.hpp]: move this (and others) out of line.

File: rtl/eclrtl/rtlrecord.hpp [hunk 12]
```diff
@@ -25,22 +25,189 @@
 #endif
 
 #include "eclrtl_imp.hpp"
+#include "rtlds_imp.hpp"
 #include "rtlfield.hpp"
 
+//---------------------------------------------------------------------------
+// Record size handling.
+
+class CGlobalHelperClass : public RtlCInterface
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.hpp]: I suspect all these classes now need 
ECLRTL_API
added to the class definition.

File: rtl/eclrtl/rtlrecord.cpp (hunk 6 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.cpp]: fixed

File: rtl/eclrtl/rtlrecord.hpp (hunk 8 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.hpp]: Removed - nothing seems to have broken yet. It's not referenced from generated code?

File: rtl/eclrtl/rtlrecord.hpp (hunk 9 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.hpp]: Fixed

File: rtl/eclrtl/rtlrecord.hpp (hunk 12 above)
Comment (richardkchapman) [inline rtl/eclrtl/rtlrecord.hpp]: Not sure why - it has moved from one header to another is all...

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

**Comment by ghalliday** [comment]
@richardkchapman some final comments.

Also, should this contain a test case which calls your function(s).

**Pull request merged**